### PR TITLE
Spree::Payment::Processing refactor

### PR DESCRIPTION
**Pull request by elia**
## Summary

While [working on solidus_stripe](https://github.com/solidusio/solidus_stripe/issues/152) I had a deeper look at how the processing is handled and took the chance to straight out a few indirections.

Since it's only touching private APIs we shouldn't need any special handling or deprecations.

## Checklist

Check out our [PR guidelines](https://github.com/solidusio/.github/blob/master/CONTRIBUTING.md#pull-request-guidelines) for more details.

The following are mandatory for all PRs:

- [x] I have written a thorough PR description.
- [x] I have kept my commits small and atomic.
- [x] I have used clear, explanatory commit messages.

The following are not always needed (~cross them out~ if they are not):

<del>
- [ ] I have added automated tests to cover my changes.
- [ ] I have attached screenshots to demo visual changes.
- [ ] I have opened a PR to update the [guides](https://github.com/solidusio/edgeguides).
- [ ] I have updated the README to account for my changes.
</del>

### CODE REVIEW

**Review by jarednorman — APPROVED** [review]

**Review by adammathys — APPROVED** [review]

**Review by tvdeyen — APPROVED** [review]
Thanks for clearing this up! 🙏🏻

File: core/app/models/spree/payment/processing.rb [hunk 1]
```diff
@@ -35,12 +35,37 @@ def process!
       end
 
       def authorize!
-        handle_payment_preconditions { process_authorization }
```
Comment (kennyadsl) [inline core/app/models/spree/payment/processing.rb:38]: - https://github.com/search?q=org%3Asolidusio+handle_payment_preconditions&type=code
- https://github.com/search?q=org%3Asolidusio-contrib+handle_payment_preconditions&type=code

Not used in any extension 🎉

File: core/app/models/spree/payment/processing.rb [hunk 2]
```diff
@@ -35,12 +35,37 @@ def process!
...
+        protect_from_connection_error do
+          response = payment_method.purchase(
+            money.money.cents,
+            source,
+            gateway_options,
+          )
+          complete! if handle_response(response)
+        end
+
+        # This won't be called if gateway_action raises a GatewayError
```
Comment (kennyadsl) [inline core/app/models/spree/payment/processing.rb:67]: We removed `gateway_action`, should we also change this comment?

File: core/app/models/spree/payment/processing.rb [hunk 3]
```diff
@@ -139,77 +164,56 @@ def handle_void_response(response)
...
-          end
-        else
-          raise Core::GatewayError.new(I18n.t('spree.payment_processing_failed'))
+        unless payment_method.supports?(source)
+          invalidate!
+          gateway_error(I18n.t('spree.payment_method_not_supported'))
         end
-      end
 
-      def gateway_action(source, action, success_state)
```
Comment (kennyadsl) [inline core/app/models/spree/payment/processing.rb:176]: Found an occurrence of this here: https://github.com/solidusio/solidus_paypal_braintree/blob/806d20b50fbb2af6762b3dabc6a210306fa7d89f/spec/models/solidus_paypal_braintree/gateway_spec.rb#L103. Maybe we can change the spec first to stop relying on a private method if possible?

**Review by kennyadsl — COMMENTED** [review]
Left some comments here and there but I think that this change is worth it, there was a lot of premature optimization in this class, while we probably need more simplicity.

File: core/app/models/spree/payment/processing.rb (hunk 3 above)
Comment (elia) [reply core/app/models/spree/payment/processing.rb:176]: Good finding!

Luckily this looks like a false positive, `gateway_action` in that contexts refers to an RSpec `let` that points to a method of `Braintree::TransactionGateway`.

File: core/app/models/spree/payment/processing.rb (hunk 2 above)
Comment (elia) [reply core/app/models/spree/payment/processing.rb]: Updated 👍

File: core/app/models/spree/user_last_url_storer.rb [hunk 4]
```diff
@@ -26,7 +26,7 @@ def self.rules
     #  or its subclasses. The controller will be passed to each rule for matching.
     def initialize(controller)
       @controller = controller
-      Spree::Deprecation.warn("This class will be removed without replacement on the release of Solidus 4.0")
+      Spree::Deprecation.warn("Spree::UserLastUrlStorer will be removed without replacement on the release of Solidus 4.0")
```
[the remark this answers is not in the document]
Comment (elia) [reply core/app/models/spree/user_last_url_storer.rb]: Not an issue ☺️ 
here it is https://github.com/solidusio/solidus/pull/4838

File: core/app/models/spree/payment/processing.rb [hunk 5]
```diff
@@ -35,12 +35,37 @@ def process!
...
+        protect_from_connection_error do
+          response = payment_method.purchase(
+            money.money.cents,
+            source,
+            gateway_options,
+          )
+          complete! if handle_response(response)
+        end
+
+        # This won't be called if `payment_method.purchase` raises a GatewayError
```
Comment (kennyadsl) [inline core/app/models/spree/payment/processing.rb:67]: Reading it again now, is this comment needed? I mean I don't expect this line to be executed if the method called before this line raises an exception. WDYT?

**Comment by kennyadsl** [comment]
@elia I think you pushed the wrong branch here. Can you please check?

**Comment by elia** [comment]
@kennyadsl 🤦‍♂️ on it thanks

**Comment by elia** [comment]
@kennyadsl done 👍

**Comment by kennyadsl** [comment]
@elia please rebase!

**Review by waiting-for-dev — APPROVED** [review]
Very nice job ❤️

File: core/app/models/spree/payment/processing.rb (hunk 5 above)
Comment (kennyadsl) [reply core/app/models/spree/payment/processing.rb:67]: @elia please one last feedback here and we are ready to merge!

Comment (elia) [reply core/app/models/spree/payment/processing.rb:67]: Good point, ruby 1:1 😅

**Pull request merged**